File: src/lib/data/layers.ts
import type { AccessibilityProfileType } from "@/types";

export type LayerKind =
  | "guiding_block"
  | "pedestrian_crossing"
  | "audio_crossing_signal"
  | "obstacle"
  | "surface_hazard"
  | "ramp"
  | "stairs"
  | "elevator"
  | "path_width"
  | "surface_condition"
  | "accessible_entrance";

export type FeatureTone = "success" | "warning" | "danger" | "neutral";

export interface LayerStatus {
  symbol: string;
  label: string;
  tone: FeatureTone;
}

export interface LayerMeta {
  label: string;
  profile: AccessibilityProfileType;
  statuses: Record<string, LayerStatus>;
}

export const LAYERS: Record<LayerKind, LayerMeta> = {
  guiding_block: {
    label: "Guiding Block",
    profile: "VISUAL_NAVIGATION",
    statuses: {
      available: { symbol: "▮", label: "Tersedia", tone: "success" },
      damaged: { symbol: "▮⌁", label: "Rusak", tone: "danger" },
      interrupted: { symbol: "- - -", label: "Terputus", tone: "warning" },
      unknown: { symbol: "?", label: "Tidak diketahui", tone: "neutral" },
    },
  },
  pedestrian_crossing: {
    label: "Zebra Crossing",
    profile: "VISUAL_NAVIGATION",
    statuses: {
      available: { symbol: "≡", label: "Ada", tone: "success" },
      signalized: { symbol: "●", label: "Ada dengan lampu", tone: "success" },
      non_signalized: { symbol: "○", label: "Tanpa lampu", tone: "warning" },
      unknown: { symbol: "?", label: "Tidak diketahui", tone: "neutral" },
    },
  },
  audio_crossing_signal: {
    label: "Sinyal Suara Penyeberangan",
    profile: "VISUAL_NAVIGATION",
    statuses: {
      available: { symbol: "🔊", label: "Tersedia", tone: "success" },
      unavailable: { symbol: "✕", label: "Tidak tersedia", tone: "danger" },
      unknown: { symbol: "?", label: "Tidak diketahui", tone: "neutral" },
    },
  },
  obstacle: {
    label: "Hambatan",
    profile: "VISUAL_NAVIGATION",
    statuses: {
      construction: { symbol: "🚧", label: "Konstruksi", tone: "danger" },
      permanent: { symbol: "⛔", label: "Permanen", tone: "danger" },
      temporary: { symbol: "🟨", label: "Sementara", tone: "warning" },
      street_furniture: { symbol: "◎", label: "Fasilitas jalan", tone: "warning" },
      other: { symbol: "⚠", label: "Lainnya", tone: "warning" },
    },
  },
  surface_hazard: {
    label: "Bahaya Permukaan",
    profile: "VISUAL_NAVIGATION",
    statuses: {
      hole: { symbol: "◌", label: "Lubang", tone: "danger" },
      damaged_sidewalk: { symbol: "▥", label: "Trotoar rusak", tone: "danger" },
      uneven_surface: { symbol: "⤫", label: "Permukaan tidak rata", tone: "warning" },
      other: { symbol: "⚠", label: "Lainnya", tone: "warning" },
    },
  },
  ramp: {
    label: "Ramp",
    profile: "WHEELCHAIR_MOBILITY",
    statuses: {
      available: { symbol: "↘", label: "Tersedia", tone: "success" },
      damaged: { symbol: "↘⚠", label: "Rusak", tone: "danger" },
      unknown: { symbol: "?", label: "Tidak diketahui", tone: "neutral" },
    },
  },
  stairs: {
    label: "Tangga",
    profile: "WHEELCHAIR_MOBILITY",
    statuses: {
      none: { symbol: "✓", label: "Tanpa tangga", tone: "success" },
      present: { symbol: "≡", label: "Ada tangga", tone: "danger" },
      unknown: { symbol: "?", label: "Tidak diketahui", tone: "neutral" },
    },
  },
  elevator: {
    label: "Elevator",
    profile: "WHEELCHAIR_MOBILITY",
    statuses: {
      available: { symbol: "↕", label: "Tersedia", tone: "success" },
      unavailable: { symbol: "✕", label: "Tidak tersedia", tone: "danger" },
      out_of_service: { symbol: "🚧", label: "Sedang diperbaiki", tone: "danger" },
      unknown: { symbol: "?", label: "Tidak diketahui", tone: "neutral" },
    },
  },
  path_width: {
    label: "Lebar Jalur",
    profile: "WHEELCHAIR_MOBILITY",
    statuses: {
      accessible: { symbol: "⇔", label: "Aksesibel", tone: "success" },
      limited: { symbol: "↔", label: "Terbatas", tone: "warning" },
      too_narrow: { symbol: "╳", label: "Terlalu sempit", tone: "danger" },
      unknown: { symbol: "?", label: "Tidak diketahui", tone: "neutral" },
    },
  },
  surface_condition: {
    label: "Kondisi Permukaan",
    profile: "WHEELCHAIR_MOBILITY",
    statuses: {
      good: { symbol: "✓", label: "Baik", tone: "success" },
      uneven: { symbol: "⤫", label: "Tidak rata", tone: "warning" },
      damaged: { symbol: "▥", label: "Rusak", tone: "danger" },
      blocked: { symbol: "✕", label: "Terhalang", tone: "danger" },
    },
  },
  accessible_entrance: {
    label: "Akses Masuk",
    profile: "WHEELCHAIR_MOBILITY",
    statuses: {
      accessible: { symbol: "↘", label: "Aksesibel", tone: "success" },
      partially_accessible: { symbol: "⚠", label: "Sebagian aksesibel", tone: "warning" },
      not_accessible: { symbol: "✕", label: "Tidak aksesibel", tone: "danger" },
      unknown: { symbol: "?", label: "Tidak diketahui", tone: "neutral" },
    },
  },
};

export const LAYER_KINDS = Object.keys(LAYERS) as LayerKind[];

export function layersForProfile(profile: AccessibilityProfileType | null): LayerKind[] {
  if (!profile) return LAYER_KINDS;
  return LAYER_KINDS.filter((kind) => LAYERS[kind].profile === profile);
}

export function statusMeta(kind: LayerKind, status: string): LayerStatus {
  return LAYERS[kind].statuses[status] ?? { symbol: "?", label: "Tidak diketahui", tone: "neutral" };
}

File: backend/internal/model/layers.go
package model

// LayerStatus is the display metadata for a single feature kind/status pair.
type LayerStatus struct {
	Symbol string
	Label  string
	Tone   string
}

// LayerMeta describes an accessibility feature layer.
type LayerMeta struct {
	Label    string
	Profile  string
	Statuses map[string]LayerStatus
}

// LAYERS is the catalog of accessibility map layers. Profiles:
// VISUAL_NAVIGATION or WHEELCHAIR_MOBILITY.
var LAYERS = map[string]*LayerMeta{
	"guiding_block": {
		Label: "Guiding Block", Profile: "VISUAL_NAVIGATION",
		Statuses: map[string]LayerStatus{
			"available":   {Symbol: "▮", Label: "Tersedia", Tone: "success"},
			"damaged":     {Symbol: "▮⌁", Label: "Rusak", Tone: "danger"},
			"interrupted": {Symbol: "- - -", Label: "Terputus", Tone: "warning"},
			"unknown":     {Symbol: "?", Label: "Tidak diketahui", Tone: "neutral"},
		},
	},
	"pedestrian_crossing": {
		Label: "Zebra Crossing", Profile: "VISUAL_NAVIGATION",
		Statuses: map[string]LayerStatus{
			"available":      {Symbol: "≡", Label: "Ada", Tone: "success"},
			"signalized":     {Symbol: "●", Label: "Ada dengan lampu", Tone: "success"},
			"non_signalized": {Symbol: "○", Label: "Tanpa lampu", Tone: "warning"},
			"unknown":        {Symbol: "?", Label: "Tidak diketahui", Tone: "neutral"},
		},
	},
	"audio_crossing_signal": {
		Label: "Sinyal Suara Penyeberangan", Profile: "VISUAL_NAVIGATION",
		Statuses: map[string]LayerStatus{
			"available":   {Symbol: "🔊", Label: "Tersedia", Tone: "success"},
			"unavailable": {Symbol: "✕", Label: "Tidak tersedia", Tone: "danger"},
			"unknown":     {Symbol: "?", Label: "Tidak diketahui", Tone: "neutral"},
		},
	},
	"obstacle": {
		Label: "Hambatan", Profile: "VISUAL_NAVIGATION",
		Statuses: map[string]LayerStatus{
			"construction":     {Symbol: "🚧", Label: "Konstruksi", Tone: "danger"},
			"permanent":        {Symbol: "⛔", Label: "Permanen", Tone: "danger"},
			"temporary":        {Symbol: "🟨", Label: "Sementara", Tone: "warning"},
			"street_furniture": {Symbol: "◎", Label: "Fasilitas jalan", Tone: "warning"},
			"other":            {Symbol: "⚠", Label: "Lainnya", Tone: "warning"},
		},
	},
	"surface_hazard": {
		Label: "Bahaya Permukaan", Profile: "VISUAL_NAVIGATION",
		Statuses: map[string]LayerStatus{
			"hole":             {Symbol: "◌", Label: "Lubang", Tone: "danger"},
			"damaged_sidewalk": {Symbol: "▥", Label: "Trotoar rusak", Tone: "danger"},
			"uneven_surface":   {Symbol: "⤫", Label: "Permukaan tidak rata", Tone: "warning"},
			"other":            {Symbol: "⚠", Label: "Lainnya", Tone: "warning"},
		},
	},
	"ramp": {
		Label: "Ramp", Profile: "WHEELCHAIR_MOBILITY",
		Statuses: map[string]LayerStatus{
			"available": {Symbol: "↘", Label: "Tersedia", Tone: "success"},
			"damaged":   {Symbol: "↘⚠", Label: "Rusak", Tone: "danger"},
			"unknown":   {Symbol: "?", Label: "Tidak diketahui", Tone: "neutral"},
		},
	},
	"stairs": {
		Label: "Tangga", Profile: "WHEELCHAIR_MOBILITY",
		Statuses: map[string]LayerStatus{
			"none":    {Symbol: "✓", Label: "Tanpa tangga", Tone: "success"},
			"present": {Symbol: "≡", Label: "Ada tangga", Tone: "danger"},
			"unknown": {Symbol: "?", Label: "Tidak diketahui", Tone: "neutral"},
		},
	},
	"elevator": {
		Label: "Elevator", Profile: "WHEELCHAIR_MOBILITY",
		Statuses: map[string]LayerStatus{
			"available":      {Symbol: "↕", Label: "Tersedia", Tone: "success"},
			"unavailable":    {Symbol: "✕", Label: "Tidak tersedia", Tone: "danger"},
			"out_of_service": {Symbol: "🚧", Label: "Sedang diperbaiki", Tone: "danger"},
			"unknown":        {Symbol: "?", Label: "Tidak diketahui", Tone: "neutral"},
		},
	},
	"path_width": {
		Label: "Lebar Jalur", Profile: "WHEELCHAIR_MOBILITY",
		Statuses: map[string]LayerStatus{
			"accessible": {Symbol: "⇔", Label: "Aksesibel", Tone: "success"},
			"limited":    {Symbol: "↔", Label: "Terbatas", Tone: "warning"},
			"too_narrow": {Symbol: "╳", Label: "Terlalu sempit", Tone: "danger"},
			"unknown":    {Symbol: "?", Label: "Tidak diketahui", Tone: "neutral"},
		},
	},
	"surface_condition": {
		Label: "Kondisi Permukaan", Profile: "WHEELCHAIR_MOBILITY",
		Statuses: map[string]LayerStatus{
			"good":    {Symbol: "✓", Label: "Baik", Tone: "success"},
			"uneven":  {Symbol: "⤫", Label: "Tidak rata", Tone: "warning"},
			"damaged": {Symbol: "▥", Label: "Rusak", Tone: "danger"},
			"blocked": {Symbol: "✕", Label: "Terhalang", Tone: "danger"},
		},
	},
	"accessible_entrance": {
		Label: "Akses Masuk", Profile: "WHEELCHAIR_MOBILITY",
		Statuses: map[string]LayerStatus{
			"accessible":           {Symbol: "↘", Label: "Aksesibel", Tone: "success"},
			"partially_accessible": {Symbol: "⚠", Label: "Sebagian aksesibel", Tone: "warning"},
			"not_accessible":       {Symbol: "✕", Label: "Tidak aksesibel", Tone: "danger"},
			"unknown":              {Symbol: "?", Label: "Tidak diketahui", Tone: "neutral"},
		},
	},
}

// LayersForProfile returns the feature kinds visible for a profile (or all
// when profile is empty).
func LayersForProfile(profile string) []string {
	var out []string
	for kind, meta := range LAYERS {
		if profile == "" || meta.Profile == profile {
			out = append(out, kind)
		}
	}
	return out
}

// StatusMeta returns the display metadata for a feature kind/status pair.
func StatusMeta(kind, status string) LayerStatus {
	if meta, ok := LAYERS[kind]; ok {
		if s, ok := meta.Statuses[status]; ok {
			return s
		}
	}
	return LayerStatus{Symbol: "?", Label: "Tidak diketahui", Tone: "neutral"}
}

// ReportCategory is a reportable accessibility issue.
type ReportCategory struct {
	Value string
	Label string
}

// REPORT_CATEGORIES lists the selectable report categories.
var REPORT_CATEGORIES = []ReportCategory{
	{"STAIRS", "Tangga (tanpa ramp/tangan)"},
	{"DAMAGED_RAMP", "Ramp rusak"},
	{"RAMP", "Ramp tersedia"},
	{"GUIDING_BLOCK", "Guiding block hilang/terputus"},
	{"DAMAGED_SIDEWALK", "Trotoar rusak"},
	{"OBSTACLE", "Hambatan di jalur"},
	{"ELEVATOR", "Elevator tidak berfungsi"},
	{"ACCESSIBLE_FACILITY", "Fasilitas aksesibel tersedia"},
	{"OTHER", "Lainnya"},
}

// ReportCategoryLabel returns the human label for a category value.
func ReportCategoryLabel(category string) string {
	for _, c := range REPORT_CATEGORIES {
		if c.Value == category {
			return c.Label
		}
	}
	return "Lainnya"
}

// IsReportCategory reports whether value is a known category.
func IsReportCategory(value string) bool {
	for _, c := range REPORT_CATEGORIES {
		if c.Value == value {
			return true
		}
	}
	return false
}

// REPORT_STATUS_META maps a report status to display metadata.
var REPORT_STATUS_META = map[string]struct{ Label, Tone, Symbol string }{
	"PENDING":  {Label: "Menunggu verifikasi", Tone: "neutral", Symbol: "⌛"},
	"VERIFIED": {Label: "Terverifikasi", Tone: "success", Symbol: "✓"},
	"ACTIVE":   {Label: "Masih relevan", Tone: "success", Symbol: "✓"},
	"OUTDATED": {Label: "Mungkin usang", Tone: "warning", Symbol: "⚠"},
	"RESOLVED": {Label: "Sudah diperbaiki", Tone: "neutral", Symbol: "✓"},
	"REJECTED": {Label: "Ditolak", Tone: "danger", Symbol: "✕"},
}

// Known enum values.
var (
	ACCESSIBILITY_PROFILES = []string{"VISUAL_NAVIGATION", "WHEELCHAIR_MOBILITY"}
	SEVERITIES             = []string{"HIGH", "MEDIUM", "LOW"}
	STATUSES               = []string{"PENDING", "VERIFIED", "ACTIVE", "OUTDATED", "RESOLVED", "REJECTED"}
	VERIFICATION_TYPES     = []string{"CONFIRMED", "CHANGED", "RESOLVED"}
)

// IsAccessibilityProfile reports whether value is a known accessibility profile.
func IsAccessibilityProfile(value string) bool {
	for _, p := range ACCESSIBILITY_PROFILES {
		if p == value {
			return true
		}
	}
	return false
}

// IsStatus reports whether value is a known report status.
func IsStatus(s string) bool {
	for _, x := range STATUSES {
		if x == s {
			return true
		}
	}
	return false
}

// IsSeverity reports whether value is a known severity.
func IsSeverity(s string) bool {
	for _, x := range SEVERITIES {
		if x == s {
			return true
		}
	}
	return false
}

// IsVerificationType reports whether value is a known verification type.
func IsVerificationType(s string) bool {
	for _, x := range VERIFICATION_TYPES {
		if x == s {
			return true
		}
	}
	return false
}

// IsAffectedProfile reports whether value is a valid affected-profiles token.
func IsAffectedProfile(s string) bool {
	return s == "BOTH" || s == "WHEELCHAIR_MOBILITY" || s == "VISUAL_NAVIGATION"
}

// Photo upload limits.
const (
	MaxPhotoBytes             = 5 * 1024 * 1024
	MaxVerificationsPerReport = 30
	MaxPhotosPerReport        = 5
)

// ALLOWED_PHOTO_TYPES lists the MIME types accepted for report photos.
var ALLOWED_PHOTO_TYPES = []string{"image/png", "image/jpeg", "image/webp"}

// ReportStatusLabel returns the human label for a report status.
func ReportStatusLabel(status string) string {
	if m, ok := REPORT_STATUS_META[status]; ok {
		return m.Label
	}
	return status
}
